File: packages/core/src/drivers/git-driver.ts
import { mkdir, readdir } from "node:fs/promises";
import { basename } from "node:path";
import { LocalConfigError } from "../errors.js";
import { diagnoseGitAuth } from "../git-auth.js";
import type { DiagnosticCheck, GitRepositoryConfig, RepositoryStatus } from "../model.js";
import { runProcess } from "../process.js";
import type { DriverContext, GitDriverContract, PullResult, PushResult } from "../repository-driver.js";

const capabilities = { history: true, conditionalWrite: true, atomicPublish: true } as const;

function lines(value: string): string[] { return value ? value.split(/\r?\n/).filter(Boolean) : []; }
function normalizeStatusPath(line: string): string {
  const raw = line.slice(3).trim();
  const renamed = raw.includes(" -> ") ? raw.split(" -> ").at(-1)! : raw;
  return renamed.replace(/^"|"$/g, "");
}
function inScope(path: string, scopes: string[]): boolean { return scopes.some((scope) => path === scope || path.startsWith(`${scope}/`)); }

async function remoteRevision(repository: GitRepositoryConfig): Promise<string | undefined> {
  const result = await runProcess("git", ["ls-remote", repository.options.remoteUrl, `refs/heads/${repository.options.branch}`], { allowFailure: true });
  if (result.exitCode !== 0) {
    const auth = /authentication|permission denied|could not read|access denied/i.test(result.stderr);
    throw new LocalConfigError(auth ? "auth_failed" : "repository_failed", "Cannot read Git remote revision", { repositoryId: repository.id });
  }
  return result.stdout.split(/\s+/)[0] || undefined;
}

async function localRevision(repository: GitRepositoryConfig): Promise<string | undefined> {
  const result = await runProcess("git", ["rev-parse", "HEAD"], { cwd: repository.workspacePath, allowFailure: true });
  return result.exitCode === 0 ? result.stdout : undefined;
}

export class GitDriver implements GitDriverContract {
  readonly type = "git" as const;
  readonly capabilities = capabilities;

  async prepare(repository: GitRepositoryConfig): Promise<void> {
    await mkdir(repository.workspacePath, { recursive: true });
    const entries = await readdir(repository.workspacePath);
    if (entries.length === 0) {
      const clone = await runProcess("git", ["clone", "--origin", "origin", "--branch", repository.options.branch, "--single-branch", repository.options.remoteUrl, repository.workspacePath], { allowFailure: true });
      if (clone.exitCode !== 0) {
        // An empty remote has no branch yet; initialize it without inventing a remote commit.
        const emptyRemote = /empty repository|remote branch .* not found/i.test(`${clone.stdout}\n${clone.stderr}`);
        if (!emptyRemote) {
          const auth = /authentication|permission denied|could not read|access denied/i.test(clone.stderr);
          throw new LocalConfigError(auth ? "auth_failed" : "repository_failed", `Cannot clone Git repository: ${clone.stderr}`, { repositoryId: repository.id });
        }
        await runProcess("git", ["init", "--initial-branch", repository.options.branch], { cwd: repository.workspacePath });
        await runProcess("git", ["remote", "add", "origin", repository.options.remoteUrl], { cwd: repository.workspacePath });
      }
      return;
    }
    const inside = await runProcess("git", ["rev-parse", "--is-inside-work-tree"], { cwd: repository.workspacePath, allowFailure: true });
    if (inside.stdout !== "true") throw new LocalConfigError("repository_failed", "Managed Git workspace is not a Git repository", { workspacePath: repository.workspacePath });
    const url = await runProcess("git", ["remote", "get-url", "origin"], { cwd: repository.workspacePath, allowFailure: true });
    if (url.exitCode !== 0 || url.stdout !== repository.options.remoteUrl) {
      throw new LocalConfigError("repository_failed", "Managed workspace origin does not match configured remote", { configured: repository.options.remoteUrl, actual: url.stdout });
    }
  }

  async inspect(context: DriverContext<GitRepositoryConfig>): Promise<RepositoryStatus> {
    await this.prepare(context.repository);
    const [remote, local, dirty] = await Promise.all([
      remoteRevision(context.repository),
      localRevision(context.repository),
      runProcess("git", ["status", "--porcelain=v1", "--untracked-files=all"], { cwd: context.repository.workspacePath }),
    ]);
    const localChanges = lines(dirty.stdout).map(normalizeStatusPath);
    return {
      state: localChanges.length || remote !== local ? "pending" : "synced",
      remoteRevision: remote,
      localChanges,
      remoteChanged: remote !== local,
      capabilities,
    };
  }

  async pull(context: DriverContext<GitRepositoryConfig>): Promise<PullResult> {
    const before = await localRevision(context.repository);
    const remote = await remoteRevision(context.repository);
    if (!remote || remote === before) return { remoteRevision: remote ?? before, changed: false };
    const dirty = lines((await runProcess("git", ["status", "--porcelain=v1", "--untracked-files=all"], { cwd: context.repository.workspacePath })).stdout).map(normalizeStatusPath);
    if (dirty.length) {
      throw new LocalConfigError("conflict", "Remote changed while the managed workspace has local changes", { repositoryId: context.repository.id, paths: dirty });
    }
    await runProcess("git", ["fetch", "--prune", "origin", context.repository.options.branch], { cwd: context.repository.workspacePath });
    const merge = await runProcess("git", ["merge", "--ff-only", `origin/${context.repository.options.branch}`], { cwd: context.repository.workspacePath, allowFailure: true });
    if (merge.exitCode !== 0) throw new LocalConfigError("conflict", "Git history diverged; resolve it manually in the managed workspace", { repositoryId: context.repository.id, workspacePath: context.repository.workspacePath });
    return { remoteRevision: await localRevision(context.repository), changed: true };
  }

  async push(context: DriverContext<GitRepositoryConfig>, commitMessage: string): Promise<PushResult> {
    const currentRemote = await remoteRevision(context.repository);
    if (currentRemote !== context.expectedRevision) {
      throw new LocalConfigError("conflict", "Remote revision changed before push", { repositoryId: context.repository.id, expectedRevision: context.expectedRevision, remoteRevision: currentRemote });
    }
    const status = lines((await runProcess("git", ["status", "--porcelain=v1", "--untracked-files=all"], { cwd: context.repository.workspacePath })).stdout);
    const changed = status.map(normalizeStatusPath);
    const outside = changed.filter((path) => !inScope(path, context.scopes));
    if (outside.length) throw new LocalConfigError("repository_dirty_outside_scope", "Repository has changes outside the requested mapping scope", { repositoryId: context.repository.id, paths: outside });
    if (!changed.length) return { remoteRevision: currentRemote, changed: false };
    await runProcess("git", ["add", "--all", "--", ...context.scopes], { cwd: context.repository.workspacePath });
    const staged = await runProcess("git", ["diff", "--cached", "--quiet"], { cwd: context.repository.workspacePath, allowFailure: true });
    if (staged.exitCode === 0) return { remoteRevision: currentRemote, changed: false };
    const identity = await runProcess("git", ["config", "user.email"], { cwd: context.repository.workspacePath, allowFailure: true });
    if (identity.exitCode !== 0) throw new LocalConfigError("repository_failed", "Git author identity is not configured. Set user.name and user.email.", { workspacePath: context.repository.workspacePath });
    await runProcess("git", ["commit", "-m", commitMessage], { cwd: context.repository.workspacePath });
    const pushed = await runProcess("git", ["push", "origin", `HEAD:refs/heads/${context.repository.options.branch}`], { cwd: context.repository.workspacePath, allowFailure: true });
    if (pushed.exitCode !== 0) {
      const auth = /authentication|permission denied|could not read|access denied/i.test(pushed.stderr);
      const rejected = /rejected|fetch first|non-fast-forward/i.test(pushed.stderr);
      throw new LocalConfigError(auth ? "auth_failed" : rejected ? "conflict" : "repository_failed", `Git push failed: ${pushed.stderr}`, { repositoryId: context.repository.id });
    }
    return { remoteRevision: await localRevision(context.repository), changed: true };
  }

  async doctor(repository: GitRepositoryConfig) {
    const checks: DiagnosticCheck[] = [];
    const git = await runProcess("git", ["--version"], { allowFailure: true });
    checks.push({ name: "git-cli", ok: git.exitCode === 0, message: git.exitCode === 0 ? git.stdout : "Git CLI is not available", remediation: git.exitCode === 0 ? undefined : "Install Git and ensure it is on PATH" });
    if (git.exitCode === 0) checks.push(...await diagnoseGitAuth(repository));
    const usable = await runProcess("git", ["rev-parse", "--is-inside-work-tree"], { cwd: repository.workspacePath, allowFailure: true });
    checks.push({ name: "workspace", ok: usable.stdout === "true", message: usable.stdout === "true" ? "Managed workspace is valid" : `Managed workspace is not ready: ${basename(repository.workspacePath)}` });
    return { ok: checks.every((check) => check.ok || check.name === "github-cli-auth"), checks };
  }
}

File: internal/core/git_driver.go
package core

import (
	"os"
	"path/filepath"
	"regexp"
	"strings"
)

var gitCapabilities = DriverCapabilities{History: true, ConditionalWrite: true, AtomicPublish: true}
var emptyRemotePattern = regexp.MustCompile(`(?i)empty repository|remote branch .* not found`)
var authPattern = regexp.MustCompile(`(?i)authentication|permission denied|could not read|access denied`)
var rejectedPattern = regexp.MustCompile(`(?i)rejected|fetch first|non-fast-forward`)

type GitDriver struct{}

func splitLines(value string) []string {
	if value == "" {
		return []string{}
	}
	var lines []string
	for _, line := range strings.Split(strings.ReplaceAll(value, "\r\n", "\n"), "\n") {
		if line != "" {
			lines = append(lines, line)
		}
	}
	return lines
}

func normalizeStatusPath(line string) string {
	if len(line) >= 3 {
		line = strings.TrimSpace(line[3:])
	}
	if index := strings.LastIndex(line, " -> "); index >= 0 {
		line = line[index+4:]
	}
	return strings.Trim(line, `"`)
}

func inScope(path string, scopes []string) bool {
	for _, scope := range scopes {
		if path == scope || strings.HasPrefix(path, scope+"/") {
			return true
		}
	}
	return false
}

func remoteRevision(repository Repository) (string, error) {
	result, err := RunProcess("git", []string{"ls-remote", repository.Options.RemoteURL, "refs/heads/" + repository.Options.Branch}, "", nil, true)
	if err != nil {
		return "", err
	}
	if result.ExitCode != 0 {
		code := ErrRepositoryFailed
		if authPattern.MatchString(result.Stderr) {
			code = ErrAuthFailed
		}
		return "", NewError(code, "Cannot read Git remote revision", map[string]any{"repositoryId": repository.ID})
	}
	fields := strings.Fields(result.Stdout)
	if len(fields) == 0 {
		return "", nil
	}
	return fields[0], nil
}

func localRevision(repository Repository) (string, error) {
	result, err := RunProcess("git", []string{"rev-parse", "HEAD"}, repository.WorkspacePath, nil, true)
	if err != nil {
		return "", err
	}
	if result.ExitCode != 0 {
		return "", nil
	}
	return result.Stdout, nil
}

func (d *GitDriver) Prepare(repository Repository) error {
	if err := os.MkdirAll(repository.WorkspacePath, 0o755); err != nil {
		return err
	}
	entries, err := os.ReadDir(repository.WorkspacePath)
	if err != nil {
		return err
	}
	if len(entries) == 0 {
		clone, err := RunProcess("git", []string{"clone", "--origin", "origin", "--branch", repository.Options.Branch, "--single-branch", repository.Options.RemoteURL, repository.WorkspacePath}, "", nil, true)
		if err != nil {
			return err
		}
		if clone.ExitCode != 0 {
			if !emptyRemotePattern.MatchString(clone.Stdout + "\n" + clone.Stderr) {
				code := ErrRepositoryFailed
				if authPattern.MatchString(clone.Stderr) {
					code = ErrAuthFailed
				}
				return NewError(code, "Cannot clone Git repository: "+clone.Stderr, map[string]any{"repositoryId": repository.ID})
			}
			if _, err := RunProcess("git", []string{"init", "--initial-branch", repository.Options.Branch}, repository.WorkspacePath, nil, false); err != nil {
				return err
			}
			_, err = RunProcess("git", []string{"remote", "add", "origin", repository.Options.RemoteURL}, repository.WorkspacePath, nil, false)
			return err
		}
		return nil
	}
	inside, err := RunProcess("git", []string{"rev-parse", "--is-inside-work-tree"}, repository.WorkspacePath, nil, true)
	if err != nil {
		return err
	}
	if inside.Stdout != "true" {
		return NewError(ErrRepositoryFailed, "Managed Git workspace is not a Git repository", map[string]any{"workspacePath": repository.WorkspacePath})
	}
	remote, err := RunProcess("git", []string{"remote", "get-url", "origin"}, repository.WorkspacePath, nil, true)
	if err != nil {
		return err
	}
	if remote.ExitCode != 0 || remote.Stdout != repository.Options.RemoteURL {
		return NewError(ErrRepositoryFailed, "Managed workspace origin does not match configured remote", map[string]any{"configured": repository.Options.RemoteURL, "actual": remote.Stdout})
	}
	return nil
}

func (d *GitDriver) Inspect(context DriverContext) (RepositoryStatus, error) {
	if err := d.Prepare(context.Repository); err != nil {
		return RepositoryStatus{}, err
	}
	remote, err := remoteRevision(context.Repository)
	if err != nil {
		return RepositoryStatus{}, err
	}
	local, err := localRevision(context.Repository)
	if err != nil {
		return RepositoryStatus{}, err
	}
	dirty, err := RunProcess("git", []string{"status", "--porcelain=v1", "--untracked-files=all"}, context.Repository.WorkspacePath, nil, false)
	if err != nil {
		return RepositoryStatus{}, err
	}
	changes := []string{}
	for _, line := range splitLines(dirty.Stdout) {
		changes = append(changes, normalizeStatusPath(line))
	}
	state := "synced"
	if len(changes) > 0 || remote != local {
		state = "pending"
	}
	return RepositoryStatus{State: state, RemoteRevision: remote, LocalChanges: changes, RemoteChanged: remote != local, Capabilities: gitCapabilities}, nil
}

func (d *GitDriver) Pull(context DriverContext) (PullResult, error) {
	before, err := localRevision(context.Repository)
	if err != nil {
		return PullResult{}, err
	}
	remote, err := remoteRevision(context.Repository)
	if err != nil {
		return PullResult{}, err
	}
	if remote == "" || remote == before {
		if remote == "" {
			remote = before
		}
		return PullResult{RemoteRevision: remote}, nil
	}
	dirty, err := RunProcess("git", []string{"status", "--porcelain=v1", "--untracked-files=all"}, context.Repository.WorkspacePath, nil, false)
	if err != nil {
		return PullResult{}, err
	}
	paths := []string{}
	for _, line := range splitLines(dirty.Stdout) {
		paths = append(paths, normalizeStatusPath(line))
	}
	if len(paths) > 0 {
		return PullResult{}, NewError(ErrConflict, "Remote changed while the managed workspace has local changes", map[string]any{"repositoryId": context.Repository.ID, "paths": paths})
	}
	if _, err := RunProcess("git", []string{"fetch", "--prune", "origin", context.Repository.Options.Branch}, context.Repository.WorkspacePath, nil, false); err != nil {
		return PullResult{}, err
	}
	merge, err := RunProcess("git", []string{"merge", "--ff-only", "origin/" + context.Repository.Options.Branch}, context.Repository.WorkspacePath, nil, true)
	if err != nil {
		return PullResult{}, err
	}
	if merge.ExitCode != 0 {
		return PullResult{}, NewError(ErrConflict, "Git history diverged; resolve it manually in the managed workspace", map[string]any{"repositoryId": context.Repository.ID, "workspacePath": context.Repository.WorkspacePath})
	}
	revision, err := localRevision(context.Repository)
	return PullResult{RemoteRevision: revision, Changed: true}, err
}

func (d *GitDriver) Push(context DriverContext, commitMessage string) (PushResult, error) {
	currentRemote, err := remoteRevision(context.Repository)
	if err != nil {
		return PushResult{}, err
	}
	if currentRemote != context.ExpectedRevision {
		return PushResult{}, NewError(ErrConflict, "Remote revision changed before push", map[string]any{"repositoryId": context.Repository.ID, "expectedRevision": context.ExpectedRevision, "remoteRevision": currentRemote})
	}
	status, err := RunProcess("git", []string{"status", "--porcelain=v1", "--untracked-files=all"}, context.Repository.WorkspacePath, nil, false)
	if err != nil {
		return PushResult{}, err
	}
	changed, outside := []string{}, []string{}
	for _, line := range splitLines(status.Stdout) {
		path := normalizeStatusPath(line)
		changed = append(changed, path)
		if !inScope(path, context.Scopes) {
			outside = append(outside, path)
		}
	}
	if len(outside) > 0 {
		return PushResult{}, NewError(ErrRepositoryDirtyOutsideScope, "Repository has changes outside the requested mapping scope", map[string]any{"repositoryId": context.Repository.ID, "paths": outside})
	}
	if len(changed) == 0 {
		return PushResult{RemoteRevision: currentRemote}, nil
	}
	args := append([]string{"add", "--all", "--"}, context.Scopes...)
	if _, err := RunProcess("git", args, context.Repository.WorkspacePath, nil, false); err != nil {
		return PushResult{}, err
	}
	staged, err := RunProcess("git", []string{"diff", "--cached", "--quiet"}, context.Repository.WorkspacePath, nil, true)
	if err != nil {
		return PushResult{}, err
	}
	if staged.ExitCode == 0 {
		return PushResult{RemoteRevision: currentRemote}, nil
	}
	identity, err := RunProcess("git", []string{"config", "user.email"}, context.Repository.WorkspacePath, nil, true)
	if err != nil {
		return PushResult{}, err
	}
	if identity.ExitCode != 0 {
		return PushResult{}, NewError(ErrRepositoryFailed, "Git author identity is not configured. Set user.name and user.email.", map[string]any{"workspacePath": context.Repository.WorkspacePath})
	}
	if _, err := RunProcess("git", []string{"commit", "-m", commitMessage}, context.Repository.WorkspacePath, nil, false); err != nil {
		return PushResult{}, err
	}
	pushed, err := RunProcess("git", []string{"push", "origin", "HEAD:refs/heads/" + context.Repository.Options.Branch}, context.Repository.WorkspacePath, nil, true)
	if err != nil {
		return PushResult{}, err
	}
	if pushed.ExitCode != 0 {
		code := ErrRepositoryFailed
		if authPattern.MatchString(pushed.Stderr) {
			code = ErrAuthFailed
		} else if rejectedPattern.MatchString(pushed.Stderr) {
			code = ErrConflict
		}
		return PushResult{}, NewError(code, "Git push failed: "+pushed.Stderr, map[string]any{"repositoryId": context.Repository.ID})
	}
	revision, err := localRevision(context.Repository)
	return PushResult{RemoteRevision: revision, Changed: true}, err
}

func (d *GitDriver) Doctor(repository Repository) (DiagnosticResult, error) {
	checks := []DiagnosticCheck{}
	git, _ := RunProcess("git", []string{"--version"}, "", nil, true)
	if git.ExitCode == 0 {
		checks = append(checks, DiagnosticCheck{Name: "git-cli", OK: true, Message: git.Stdout})
	} else {
		checks = append(checks, DiagnosticCheck{Name: "git-cli", Message: "Git CLI is not available", Remediation: "Install Git and ensure it is on PATH"})
	}
	if git.ExitCode == 0 {
		authChecks, err := DiagnoseGitAuth(repository)
		if err != nil {
			return DiagnosticResult{}, err
		}
		checks = append(checks, authChecks...)
	}
	usable, _ := RunProcess("git", []string{"rev-parse", "--is-inside-work-tree"}, repository.WorkspacePath, nil, true)
	if usable.Stdout == "true" {
		checks = append(checks, DiagnosticCheck{Name: "workspace", OK: true, Message: "Managed workspace is valid"})
	} else {
		checks = append(checks, DiagnosticCheck{Name: "workspace", Message: "Managed workspace is not ready: " + filepath.Base(repository.WorkspacePath)})
	}
	ok := true
	for _, check := range checks {
		if !check.OK && check.Name != "github-cli-auth" {
			ok = false
		}
	}
	return DiagnosticResult{OK: ok, Checks: checks}, nil
}
